crates/pop-drink/src/error.rs

```diff
@@ -0,0 +1,154 @@
+use std::fmt::Debug;
+
+pub use drink::{
+	pallet_assets::Error as AssetsError, pallet_balances::Error as BalancesError,
+	pallet_contracts::Error as ContractsError,
+};
+pub use frame_support::sp_runtime::{
+	ArithmeticError, DispatchError, ModuleError, TokenError, TransactionalError,
+};
+use scale::{Decode, Encode};
+
+fn decode<T: Decode>(data: &[u8]) -> T {
+	T::decode(&mut &data[..]).expect("Decoding failed")
+}
+
+#[derive(Encode, Decode, Debug)]
+pub enum DrinkError<E>
+where
+	E: Decode + Encode + Debug,
+{
+	Raw(DispatchError),
+	Module(E),
+}
+
+impl<E> From<DrinkError<E>> for u32
+where
+	E: Decode + Encode + Debug,
+{
+	fn from(error: DrinkError<E>) -> Self {
+		let mut value = match error {
+			DrinkError::Raw(dispatch_error) => dispatch_error.encode(),
+			DrinkError::Module(module_error) => {
+				let mut module_error = module_error.encode();
+				module_error.insert(0, 3);
+				module_error
+			},
+		};
+		value.resize(4, 0);
+		u32::from_le_bytes(value.try_into().expect("qed, resized to 4 bytes line above"))
+	}
+}
+
+impl<E> From<u32> for DrinkError<E>
+where
+	E: Decode + Encode + Debug,
+{
+	fn from(value: u32) -> Self {
+		let encoded = value.to_le_bytes();
+		match encoded {
+			[3, module_error @ ..] => DrinkError::Module(decode(&module_error)),
+			_ => DrinkError::Raw(decode(&encoded)),
+		}
+	}
+}
+
+#[track_caller]
+pub fn assert_runtime_err_inner<T, R, E>(result: Result<R, E>, expected_error: DrinkError<T>)
```

Comment:
etc etc. 

When writing the docs take the stand of smart contract developers. They might be very unexperienced.

Reply:
Resolved https://github.com/r0gue-io/pop-drink/pull/14/commits/201a9b9eac1deb3bfcec3fea2b38d4418fb6b60a